lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -294,8 +308,29 @@ def write(chunk)
             end
           end
         end
+        if @running_cloudfunctions
+          # If the current group of entries is coming from a Cloud Functions
+          # function, the function name can be extracted from the tag.
+          match_data = @cloudfunctions_tag_regexp.match(tag)
+          if match_data
+            # Service name is set to Cloud Functions only for logs actually
+            # coming from a function.
+            @service_name = CLOUDFUNCTIONS_SERVICE
+            labels = write_log_entries_request['commonLabels']
+            labels["#{CLOUDFUNCTIONS_SERVICE}/function_name"] =
+              match_data['function_name']
+          else
+            # Other logs are considered as coming from the Container Engine
+            # service.
+            @service_name = CONTAINER_SERVICE
+          end
+        end
         arr.each do |time, record|
           next unless record.is_a?(Hash)
+          if @service_name == CLOUDFUNCTIONS_SERVICE && record.key?('log')
```

Comment:
right now, your logs look something like this, right:

 { "stream": "stdout", "log": "[D][2015-09-25T12:34:56.789Z][123456789-0]test log entry 0" }

I assume in the real world they're written as JSON and parsed by fluentd ('format json'), or similar?

if we made this { "stream": "stdout", "log": { "severity": "D", "timestamp": "2015-09-25T12:34:56.789Z", "execution_id": "123456789-0", "text": "test log entry 0" } }

Then fluentd would parse it for you and we wouldn't need the regex.

Even better would be if we could drop "log" and move all those fields to the top level; then it would handle the severity and timestamp (you'd have to rename it to 'time', and there's no nanosecond support yet but it is coming) automatically, and we could use label_map in the config file to map execution_id to a label, and the bulk of the custom code here wouldn't be necessary.